Issue report: Firmware flash failure

Hello, I am experiencing issue with the firmware on my Keeplink KP-9000-9XH-X v3.1 switch:

**Firmware Update Failure**: Whenever I attempt to flash a new firmware, the process consistently ends with a checksum error as shown below:

    Firmware upload started.......................................................
    ...............................................Checksum incorrect!
    Upload to flash done, will reset!
    Initializing Flash controller
    Detecting CPU: RTL8373N
    Software version: v0.1.0-g

After this, the device reboots back into the current version, completely ignoring the new image. I have tested this using the Firefox and chrome browser and attempted to flash both a new build of the current software and the original Keeplink factory firmware. The result is always the same.

I make some new bin file and tryed to flash switch - without sucess:


`tcp: found no unused connections.tcp: found no unused connections.tcp: found n                                                                                                                                                             o unused connections.tcp: found no unused connections.tcp: found no unused conne                                                                                                                                                             ctions.tcp: found no unused connections.Firmware upload started.................                                                                                                                                                             ................................................................................                                                                                                                                                             ................................................................................                                                                                                                                                             ................................................................................                                                                                                                                                             .......................................................................................................Checksum OK.`
```
Upload to flash done, will reset!
Initializing Flash controller
Detecting CPU: RTL8373N
Software version: v0.1.0-g
Build date: 2026-04-07 15:24:11
Hardware: keepLink KP-9000-9XHML-X V3.1
NIC reset
Setting MAC to: 06:17:64:db:5a:1f

init_switch called
CPU revision: 6868
leds_setup called
Configuring custom LED-muxes:     0x08144040      0x10349309      0x12454391      0x19616555      0x1c79d65a     0x0002181d
leds_setup done
RTL837X_PIN_MUX_0: 0x28000000
RTL837X_REG_LED_GLB_IO_EN: 0x77ffffff
RTL837X_REG_LED1_0_SET0: 0x01750041
RTL837X_REG_LED3_2_SET0: 0x00000000
RTL837X_REG_LED1_0_SET1: 0x00000175
RTL837X_REG_LED3_2_SET1: 0x00000000
RTL837X_REG_LED1_0_SET2: 0x00000000
RTL837X_REG_LED3_2_SET2: 0x00000000
RTL837X_REG_LED1_0_SET3: 0x00000000
RTL837X_REG_LED3_0_SET1: 0x00010000
RTL837X_REG_LED3_0_SET3: 0x00000000
RTL837X_LED_PORT_SET_SEL: 0x00010000
RTL837X_REG_LED_GLB_MUX_1: 0x08144040
RTL837X_REG_LED_GLB_MUX_2: 0x10349309
RTL837X_REG_LED_GLB_MUX_3: 0x12454391
RTL837X_REG_LED_GLB_MUX_4: 0x19616555
RTL837X_REG_LED_GLB_MUX_5: 0x1c79d65a
RTL837X_REG_LED_GLB_MUX_6: 0x0002181d
RTL837X_REG_LED_GLB_ACTIVE: 0x3ffb6dff
LED pad Configuration:
00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b
00 01 04 05 08 09 0c 09 0d 10 11 0e 14 11 12 15 15 16 18 19 1a 19 1d 1e 1c 1d 20 21
LED-set Configuration:
LED-ID          0               1               2               3
SET 0:     000041          000175          000000          000000
SET 1:     010175          000000          000000          000000
SET 2:     000000          000000          000000          000000
SET 3:     000000          000000          000000          000000
Port 0: SET 0: ( 2G5 LINK), ( 2G5 1G 100M 10M LINK ACT), (), (),
Port 1: SET 0: ( 2G5 LINK), ( 2G5 1G 100M 10M LINK ACT), (), (),
Port 2: SET 0: ( 2G5 LINK), ( 2G5 1G 100M 10M LINK ACT), (), (),
Port 3: SET 0: ( 2G5 LINK), ( 2G5 1G 100M 10M LINK ACT), (), (),
Port 4: SET 0: ( 2G5 LINK), ( 2G5 1G 100M 10M LINK ACT), (), (),
Port 5: SET 0: ( 2G5 LINK), ( 2G5 1G 100M 10M LINK ACT), (), (),
Port 6: SET 0: ( 2G5 LINK), ( 2G5 1G 100M 10M LINK ACT), (), (),
Port 7: SET 0: ( 2G5 LINK), ( 2G5 1G 100M 10M LINK ACT), (), (),
Port 8: SET 1: ( 2G5 1G 100M 10M LINK ACT 10G), (), (), (),

rtl8373_init called
  N-settingsCMD: 0xc10c-0x2000
phy_config_8224 called
RTL8224 ID: 0x82247000
R02f8-00000bed R02f4-00000bed
phy_config_8224 done

RTL837X_REG_SDS_MODES: 0x00000bff

RTL837X_REG_SDS_MODES: 0x00000bff
sds_config sds: 00, mode: 0d

RTL837X_REG_SDS_MODES: 0x00000bed
  TX_POLARITY_SWAP

rtl8224_phy_enable called
R02f8-0000000c R02f4-0000000c   N-settings
rtl8224_phy_enable done
EEE on for 00 speed 2g5
EEE on for 01 speed 2g5
EEE on for 02 speed 2g5
EEE on for 03 speed 2g5
EEE on for 04 speed 2g5
EEE on for 05 speed 2g5
EEE on for 06 speed 2g5
EEE on for 07 speed 2g5
EEE can't be enabled for SFP port 08

rtl8373_init done
Flash information:
  Manufacturer ID: 0xef
  Memory Type:     0x40
  Capacity:        0x15 = 2 MB
2 MB flash size detected. (1 MB is needed for image updating)
Checking for update image in flash... no update image found.

vlan_setup called
vlan_setup done

port_l2_setup called

port_l2_forget called
port_l2_forget done

port_l2_setup done
igmp_setup called
bandwidth_setup called
RTL837X_IGBW_CTRL: 0x0000003f
RTL837X_EGBW_CTRL: 0x00000001
bandwidth_setup done
uip_arp_init called
Configuring I2C for SFP idx=00 SCL=03, SDA=04
GPIO MUX setup not implemented for pin=00

A minimal prompt to explore the RTL8372:

Clock register: 0x00001101
Register 0x7b20/RTL837X_REG_SDS_MODES: 0x00000bed
Verifying PHY settings:

 Port   State   Link    TxGood          TxBad           RxGood          RxBad
1       On      Down    0x00000000      0x00000000      0x00000000      0x00000000
2       On      Down    0x00000000      0x00000000      0x00000000      0x00000000
3       On      Down    0x00000000      0x00000000      0x00000000      0x00000000
4       On      Down    0x00000000      0x00000000      0x00000000      0x00000000
5       On      Down    0x00000000      0x00000000      0x00000000      0x00000000
6       On      2.5G    0x00000040      0x00000000      0x00000022      0x00000000
7       On      1000M   0x00000024      0x00000000      0x0000001c      0x00000000
8       On      2.5G    0x00000036      0x00000000      0x0000002d      0x00000000
9       NO SFP  Down    0x00000000      0x00000000      0x00000000      0x00000000
GPIO MUX setup not implemented for pin=30
Setting ip: 192.168.10.247
Setting gw: 192.168.10.1
Setting netmask: 255.255.255.0

>
<new link: 2052500000, was 0000000000>

<SFP-RX OK>  Slot: 1
```



Here is log from compile:

```
studia@pc123:~/RTLPlayground$ make
mkdir -p output
mkdir -p output/uip
mkdir -p output/httpd
make -C tools
make[1]: Entering directory '/home/studia/RTLPlayground/tools'
mkdir -p output
make[1]: Leaving directory '/home/studia/RTLPlayground/tools'
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/rtlplayground.rel -c rtlplayground.c
rtlplayground.c:1640: warning 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
rtlplayground.c:1646: warning 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/rtl837x_flash.rel -c rtl837x_flash.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/rtl837x_leds.rel -c rtl837x_leds.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/rtl837x_phy.rel -c rtl837x_phy.c
rtl837x_phy.c:385: warning 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
rtl837x_phy.c:411: warning 126: unreachable code
rtl837x_phy.c:494: warning 238: inline function 'rtl8224_read_reg_u16' might lose function attributes
rtl837x_phy.c:501: warning 158: overflow in implicit constant conversion
rtl837x_phy.c:513: warning 238: inline function 'rtl8224_write_reg_u16' might lose function attributes
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/rtl837x_port.rel -c rtl837x_port.c
rtl837x_port.c:725: warning 85: in function port_ingress_vlan_filter_set unreferenced function argument : 'enabled'
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/cmd_parser.rel -c cmd_parser.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/html_data.rel -c html_data.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/rtl837x_igmp.rel -c rtl837x_igmp.c
rtl837x_igmp.c:93: warning 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
rtl837x_igmp.c:94: warning 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/rtl837x_stp.rel -c rtl837x_stp.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/rtl837x_pins.rel -c rtl837x_pins.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/dhcp.rel -c dhcp.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/machine.rel -c machine.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/cmd_editor.rel -c cmd_editor.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/rtl837x_bandwidth.rel -c rtl837x_bandwidth.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/rtl837x_init.rel -c rtl837x_init.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/uip/uip.rel -c uip/uip.c
uip/uip.c:377: warning 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
uip/uip.c:380: warning 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
uip/uip.c:1464: warning 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/uip/uip_arp.rel -c uip/uip_arp.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/uip/uiplib.rel -c uip/uiplib.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/uip/uip-fw.rel -c uip/uip-fw.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/uip/uip-neighbor.rel -c uip/uip-neighbor.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/uip/uip-split.rel -c uip/uip-split.c
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/httpd/httpd.rel -c httpd/httpd.c
httpd/httpd.c:423: warning 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
httpd/httpd.c:557: warning 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
sdcc -MMD -mmcs51 -I. -Ihttpd -Iuip -o output/httpd/page_impl.rel -c httpd/page_impl.c
sdcc -mmcs51 -I. -Ihttpd -Iuip -Wl-bHOME=0x00000 -Wl-bBANK1=0x14000 -Wl-bBANK2=0x24000 -Wl-r -o output/rtlplayground.ihx output/rtlplayground.rel output/rtl837x_flash.rel output/rtl837x_leds.rel output/rtl837x_phy.rel output/rtl837x_port.rel output/cmd_parser.rel output/html_data.rel output/rtl837x_igmp.rel output/rtl837x_stp.rel output/rtl837x_pins.rel output/dhcp.rel output/machine.rel output/cmd_editor.rel output/rtl837x_bandwidth.rel output/rtl837x_init.rel output/uip/timer.rel output/uip/uip.rel output/uip/uip_arp.rel output/uip/uiplib.rel output/uip/uip-fw.rel output/uip/uip-neighbor.rel output/uip/uip-split.rel output/httpd/httpd.rel output/httpd/page_impl.rel output/crtstart.rel output/crc16.rel
objcopy --input-target=ihex -O binary output/rtlplayground.ihx output/rtlplayground.img
if [ -e output/rtlplayground.bin ]; then rm output/rtlplayground.bin; fi
tools/output/imagebuilder -i output/rtlplayground.img output/rtlplayground.bin
Input file size: 181228
Input file contains 3 banks
Bank 0: Bytes read: 16384
Bank 1: bytes read: 49152
Bank 2: bytes read: 33772
tools/output/fileadder -a 454656 -s 524288 -d config.txt output/rtlplayground.bin
Input file size: 196608
Bytes read: 196608
Data inserted at 0x6f000, size: 56
tools/output/fileadder -a 458752 -s 524288 -d config.txt output/rtlplayground.bin
Input file size: 524288
Bytes read: 524288
Data inserted at 0x70000, size: 56
tools/output/fileadder -a 262144 -s 524288 -d html -p html_data -b BANK1 output/rtlplayground.bin
Input file size: 524288
Bytes read: 524288
Adding entries in directory html
Data inserted from html/port.svg at 0x40000, size: 1766
Data inserted from html/ports.html at 0x406e5, size: 787
Data inserted from html/stat.html at 0x409f7, size: 1292
Data inserted from html/main_info.js at 0x40f02, size: 857
Data inserted from html/eee_sub.js at 0x4125a, size: 382
Data inserted from html/switch.svg at 0x413d7, size: 1907
Data inserted from html/index.html at 0x41b49, size: 700
Data inserted from html/vlan.js at 0x41e04, size: 2667
Data inserted from html/system.js at 0x4286e, size: 3647
Data inserted from html/bandwidth.html at 0x436ac, size: 733
Data inserted from html/main.js at 0x43988, size: 8760
Data inserted from html/config.js at 0x45bbf, size: 1639
Data inserted from html/update.html at 0x46225, size: 674
Data inserted from html/bandwidth.js at 0x464c6, size: 5292
Data inserted from html/l2.js at 0x47971, size: 4427
Data inserted from html/lag.html at 0x48abb, size: 1187
Data inserted from html/stat.js at 0x48f5d, size: 6909
Data inserted from html/mirror.html at 0x4aa59, size: 1042
Data inserted from html/vlan.html at 0x4ae6a, size: 1621
Data inserted from html/favicon.ico at 0x4b4be, size: 1907
Data inserted from html/vlan_sub.js at 0x4bc30, size: 913
Data inserted from html/system.html at 0x4bfc0, size: 3557
Data inserted from html/ports.js at 0x4cda4, size: 5073
Data inserted from html/eee.js at 0x4e174, size: 2084
Data inserted from html/eee.html at 0x4e997, size: 999
Data inserted from html/lag.js at 0x4ed7d, size: 2224
Data inserted from html/mirror_sub.js at 0x4f62c, size: 1158
Data inserted from html/login.html at 0x4fab1, size: 801
Data inserted from html/style.css at 0x4fdd1, size: 3689
Data inserted from html/navigation.js at 0x50c39, size: 629
Data inserted from html/l2.html at 0x50ead, size: 558
Data inserted from html/mirror.js at 0x510da, size: 1914
Data inserted from html/sfp.svg at 0x51853, size: 1829
tools/output/crc_calculator -u output/rtlplayground.bin
Input file size: 524288
Bytes read: 524288
CRC16 is: 0xe307
Setting CRC bytes at position: 0x7fffe to CRC 1s complement 0x1cf8
```

--- a/httpd/httpd.c
+++ b/httpd/httpd.c
@@ -14,9 +14,6 @@
 #define SESSION_ID_LENGTH 12
 #define SESSION_TIMEOUT 200
 
-// SPI FLASH MEMORY PAGE SIZE.
-#define FLASHMEM_PAGE_SIZE 0x100
-
 #define CMARK_S 6
 
 #pragma codeseg BANK1
@@ -329,7 +326,7 @@ uint8_t stream_upload(uint16_t bptr)
 			}
 			crc16(p + bptr);
 			flash_buf[write_len++] = p[bptr++];
-			if (write_len >= FLASHMEM_PAGE_SIZE) {
+			if (write_len >= FLASH_PAGE_SIZE) {
 				dbg_string("len: "); dbg_short(write_len); dbg_char(' ');
 				dbg_string("CRC16: "); dbg_short(crc_value); dbg_char('\n');
 				if (uptr % FLASH_SECTOR_SIZE == 0) {
@@ -337,14 +334,14 @@ uint8_t stream_upload(uint16_t bptr)
 					flash_sector_erase();
 				}
 				flash_region.addr = uptr;
-				flash_region.len = FLASHMEM_PAGE_SIZE;
+				flash_region.len = FLASH_PAGE_SIZE;
 				flash_write_bytes(flash_buf);
-				uptr += FLASHMEM_PAGE_SIZE;
-				write_len -= FLASHMEM_PAGE_SIZE;
+				uptr += FLASH_PAGE_SIZE;
+				write_len -= FLASH_PAGE_SIZE;
 
 				// Copy the remaining byte for the next page to the beginning of the buffer.
 				if (write_len > 0) {
-					memcpy(flash_buf, flash_buf + FLASHMEM_PAGE_SIZE, write_len);
+					memcpy(flash_buf, flash_buf + FLASH_PAGE_SIZE, write_len);
 				}
 			}
 			bindex = 0;

--- a/httpd/httpd.c
+++ b/httpd/httpd.c
@@ -332,6 +332,10 @@ uint8_t stream_upload(uint16_t bptr)
 			if (write_len >= FLASHMEM_PAGE_SIZE) {
 				dbg_string("len: "); dbg_short(write_len); dbg_char(' ');
 				dbg_string("CRC16: "); dbg_short(crc_value); dbg_char('\n');
+				if (uptr % FLASH_SECTOR_SIZE == 0) {
+					flash_region.addr = uptr;
+					flash_sector_erase();
+				}
 				flash_region.addr = uptr;
 				flash_region.len = FLASHMEM_PAGE_SIZE;
 				flash_write_bytes(flash_buf);

--- a/rtl837x_flash.h
+++ b/rtl837x_flash.h
@@ -1,6 +1,8 @@
 #ifndef _RTL837X_FLASH_H_
 #define _RTL837X_FLASH_H_
 
+#define FLASH_SECTOR_SIZE 0x1000
+
 void flash_init(uint8_t enable_dio);
 void flash_read_uid(void);
 void flash_write_enable(void);

--- a/rtl837x_flash.h
+++ b/rtl837x_flash.h
@@ -1,7 +1,15 @@
 #ifndef _RTL837X_FLASH_H_
 #define _RTL837X_FLASH_H_
 
+// SPI FLASH MEMORY PAGE SIZE.
+#define FLASH_PAGE_SIZE 0x100
+// SPI FLASH MEMORY SECTOR SIZE = ERASE SIZE.
 #define FLASH_SECTOR_SIZE 0x1000
+
+#if (FLASH_SECTOR_SIZE % FLASH_PAGE_SIZE) != 0
+#error "FLASH_SECTOR_SIZE must be a multiple of FLASH_PAGE_SIZE"
+#endif
+
 
 void flash_init(uint8_t enable_dio);
 void flash_read_uid(void);

